app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx

```diff
@@ -0,0 +1,250 @@
+        onGeoAreaOptionsChange,
+        onChange,
+    } = props;
+
+    const [selectedRegion, setSelectedRegion] = useState<string>();
+    const [selectedAdminLevel, setSelectedAdminLevel] = useState<AdminLevel>();
+
+    const [
+        getGeoAreas,
+    ] = useLazyQuery<ProjectGeoAreasQuery, ProjectGeoAreasQueryVariables>(
+        GEOAREAS,
+        {
+            onCompleted: (data) => {
+                onGeoAreaOptionsChange(data.project?.geoAreas?.results);
+            },
+        },
+    );
+
+    const variables = useMemo(() => ({
+        projectId,
+    }
+    ), [projectId]);
+
+    const {
+        data: projectRegions,
+        loading: projectRegionsPending,
+    } = useQuery<ProjectRegionsQuery, ProjectRegionsQueryVariables>(
+        PROJECT_REGIONS,
+        {
+            skip: !projectId,
+            variables,
+            onCompleted: (data) => {
+                const [topRegion] = data.project?.regions ?? [];
+                const [topAdminLevel] = topRegion.adminLevels ?? [];
+                setSelectedRegion(topRegion.id);
+                setSelectedAdminLevel(topAdminLevel);
+            },
+        },
+    );
+
+    const adminLevels = projectRegions?.project?.regions
+        ?.filter((v) => v.id === selectedRegion)
+        .map((r) => r.adminLevels)
+        .flat()
+        .filter(isDefined);
+
+    const handleGeoAreasMapSelection = useCallback((values: string[]) => {
+        onChange(values);
+        getGeoAreas({
+            variables: {
+                ids: values,
+                projectId,
+            },
+        });
+    }, [onChange, getGeoAreas, projectId]);
+
+    const handleRegionChange = useCallback((newRegion: string | undefined) => {
+        setSelectedRegion(newRegion);
+        const selectedRegionDetails = projectRegions
+            ?.project?.regions?.find((region) => region.id === newRegion);
+        setSelectedAdminLevel(selectedRegionDetails?.adminLevels?.[0]);
+    }, [projectRegions]);
+
+    const handleAdminLevelChange = useCallback((value: string) => {
+        const adminLevel = adminLevels?.find((v) => v.id === value);
+        setSelectedAdminLevel(adminLevel);
+    }, [adminLevels]);
+
+    const handleRemoveItem = useCallback((value: string) => {
+        const newValues = tempGeoAreas?.filter((v) => (v !== value));
+        onChange(newValues);
+    }, [onChange, tempGeoAreas]);
+
+    const geoAreaGroupRendererParams = useCallback((key: string) => ({
+        heading: key,
+        headingSize: 'extraSmall',
+    }), []);
+
+    const geoAreasRendererParams = useCallback((_: string, geoArea: GeoArea) => ({
+        id: `${geoArea.id}`,
+        value: `${geoArea.regionTitle}/${geoArea.adminLevelTitle}/${geoArea.title}`,
+        onDismiss: handleRemoveItem,
+    }), [handleRemoveItem]);
+
+    const geoAreasList = useMemo(() => (
+        tempGeoAreas
+            ?.map((val: string) => geoAreaOptions?.find((v) => v.id === val))
+            .filter(isDefined)
```

Comment:
Let's compute this the other way round.
```javascript
geoAreasOptions.filter((item) => tempGeoAreas.includes(item))
```

The code complexity here is O(N^2), if the tempGeoAreas can be large
We should considering using a Map instead